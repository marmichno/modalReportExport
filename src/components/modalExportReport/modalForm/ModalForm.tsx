// react-hook-form
import { useForm, SubmitHandler, FormProvider } from 'react-hook-form';
// styled
import {
  StyledFormWrapper,
  StyledFormHeader,
  StyledFormText,
  StyledForm,
  StyledFormButtonsWrapper,
  StyledFormInputsWrapper,
  StyledFormButton,
  StyledError,
  StyledSuccess,
} from './StyledModalForm';
// components
import { ReportName } from './inputs/reportName/ReportName';
import { EmailTo } from './inputs/EmailTo/EmailTo';
import { Format } from './inputs/Format/Format';
import { Schedule } from './inputs/Schedule/Schedule';
import { Date } from './inputs/conditional/date/Date';
import { EverydayAt } from './inputs/conditional/everydayAt/EverydayAt';
import { Every } from './inputs/conditional/every/Every';
// hooks
import { useEffect, useState } from 'react';
// requests
import { mockPost } from '../requests/mockPost';

export interface InputForm {
  reportName: string;
  format: string;
  emailTo: string;
  schedule: string;
  repeat: {
    date: string;
    time: string;
    day: string;
  };
}

const initialPostState = {
  response: {},
  error: '',
  success: '',
};

export const ModalForm = () => {
  const methods = useForm<InputForm>({
    defaultValues: {
      reportName: '',
      format: 'excel',
      emailTo: '',
      schedule: 'noRepeat',
    },
  });
  const schedule = methods.watch('schedule');
  const [apiResponse, setApiResponse] = useState(initialPostState);

  const formSubmitHandler: SubmitHandler<InputForm> = async (data) => {
    const response = await mockPost(data);
    setApiResponse(response);
  };

  useEffect(() => {
    switch (schedule) {
      case 'noRepeat':
        methods.unregister(['repeat.date', 'repeat.day', 'repeat.time']);
        break;
      case 'specificDate':
        methods.unregister('repeat.day');
        break;
      case 'daily':
        methods.unregister(['repeat.day', 'repeat.date']);
        break;
      case 'weekly':
        methods.unregister('repeat.date');
        break;
    }
  }, [schedule, methods]);

  return (
    <StyledFormWrapper>
      <StyledFormHeader>
        <StyledFormText>Export Report</StyledFormText>
      </StyledFormHeader>

      <FormProvider {...methods}>
        <StyledForm
          onSubmit={methods.handleSubmit(formSubmitHandler)}
          noValidate
          autoComplete="off"
        >
          <StyledFormInputsWrapper>
            <ReportName />
            <Format />
            <EmailTo />
            <Schedule />
            {schedule === 'noRepeat' && null}
            {schedule === 'specificDate' && <Date />}
            {schedule === 'daily' && <EverydayAt />}
            {schedule === 'weekly' && <Every />}
          </StyledFormInputsWrapper>
          <StyledFormButtonsWrapper>
            {apiResponse.error && (
              <StyledError>{apiResponse.error}</StyledError>
            )}
            {apiResponse.success && (
              <StyledSuccess>{apiResponse.success}</StyledSuccess>
            )}
            <StyledFormButton>Cancel</StyledFormButton>
            <StyledFormButton data-testid="submitForm">OK</StyledFormButton>
          </StyledFormButtonsWrapper>
        </StyledForm>
      </FormProvider>
    </StyledFormWrapper>
  );
};
