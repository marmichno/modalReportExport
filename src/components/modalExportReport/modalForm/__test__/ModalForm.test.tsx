// axios
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
// jest
import '@testing-library/jest-dom/extend-expect';
// testing library
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
// components
import { ModalForm } from '../ModalForm';
// snapshot
import TestRenderer from 'react-test-renderer';

const mock = new MockAdapter(axios);

test('success message is visible after response 201 from api', async () => {
  render(<ModalForm />);

  const submitButton = screen.getByTestId('submitForm');

  mock.onPost('http://localhost:3000/exportReport').reply(201);

  fireEvent.click(submitButton);

  await waitFor(() => {
    expect(screen.getAllByText('data has been sent')).toBeTruthy();
  });
});

test('error message is visible after error response from api', async () => {
  render(<ModalForm />);

  const submitButton = screen.getByTestId('submitForm');

  mock.onPost('http://localhost:3000/exportReport').reply(404);

  fireEvent.click(submitButton);

  await waitFor(() => {
    expect(
      screen.getByText('Request failed with status code 404')
    ).toBeTruthy();
  });
});

test('on default schedule value no additional label is visible', async () => {
  render(<ModalForm />);

  expect(screen.queryByText('Every')).not.toBeTruthy();
  expect(screen.queryByText('Everyday at')).not.toBeTruthy();
  expect(screen.queryByText('Date')).not.toBeTruthy();
});

test('after selecting daily schedule correct label and inputs are visible', async () => {
  render(<ModalForm />);

  const scheduleDaily = screen.getByTestId('daily');

  fireEvent.click(scheduleDaily);

  expect(screen.getByText('Everyday at')).toBeTruthy();
  expect(screen.getByTestId('everyDayAtInput1')).toBeTruthy();
});

test('after selecting weekly schedule correct label and inputs are visible', async () => {
  render(<ModalForm />);

  const scheduleDaily = screen.getByTestId('weekly');

  fireEvent.click(scheduleDaily);

  expect(screen.getByText('Every')).toBeTruthy();
  expect(screen.getByTestId('everyInput1')).toBeTruthy();
  expect(screen.getByTestId('everyInput2')).toBeTruthy();
});

test('after selecting specific date schedule correct label and inputs are visible', async () => {
  render(<ModalForm />);

  const scheduleDaily = screen.getByTestId('specificDate');

  fireEvent.click(scheduleDaily);

  expect(screen.getByText('Date')).toBeTruthy();
  expect(screen.getByTestId('dateInput1')).toBeTruthy();
  expect(screen.getByTestId('dateInput2')).toBeTruthy();
});

test('matches snapshot', () => {
  const tree = TestRenderer.create(<ModalForm />).toJSON();
  expect(tree).toMatchSnapshot();
});
