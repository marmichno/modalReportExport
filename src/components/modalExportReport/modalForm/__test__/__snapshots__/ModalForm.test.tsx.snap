// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-bdvvtL HiMlu"
>
  <div
    className="sc-gsDKAQ fEsrgt"
  >
    <p
      className="sc-gKclnd gefYSt"
    >
      Export Report
    </p>
  </div>
  <form
    autoComplete="off"
    className="sc-dkPtRN bJDhMV"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="sc-hKwDye jXdvFQ"
    >
      <div
        className="sc-pVTFL feqbJT"
      >
        <div
          className="sc-jrQzAO cORcFK"
        >
          Report Name
        </div>
        <div
          className="sc-kDTinF kcBoLH"
        >
          <input
            className="sc-iqseJM gYSIeQ"
            name="reportName"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Shareablee Report"
            value=""
          />
        </div>
      </div>
      <div
        className="sc-pVTFL feqbJT"
      >
        <div
          className="sc-jrQzAO cORcFK"
        >
          Format
        </div>
        <div
          className="sc-kDTinF kcBoLH"
        >
          <div
            className="sc-bqiRlB kSktzV"
          >
            <input
              className="sc-crHmcD eUERYU"
              defaultChecked={true}
              name="Format"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="excel"
            />
            <label
              className="sc-egiyK gznvPK"
              htmlFor="Format"
            >
              Excel
            </label>
          </div>
          <div
            className="sc-bqiRlB kSktzV"
          >
            <input
              className="sc-crHmcD eUERYU"
              name="Format"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="csv"
            />
            <label
              className="sc-egiyK gznvPK"
              htmlFor="Format"
            >
              CSV
            </label>
          </div>
        </div>
      </div>
      <div
        className="sc-pVTFL feqbJT"
      >
        <div
          className="sc-jrQzAO cORcFK"
        >
          E-mail to
        </div>
        <div
          className="sc-kDTinF kcBoLH"
        >
          <input
            className="sc-iqseJM gYSIeQ"
            name="emailTo"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="client@company.com"
            value=""
          />
        </div>
      </div>
      <div
        className="sc-pVTFL feqbJT"
      >
        <div
          className="sc-jrQzAO cORcFK"
        >
          Schedule
        </div>
        <div
          className="sc-kDTinF kcBoLH"
        >
          <div
            className="sc-bqiRlB kSktzV"
          >
            <input
              className="sc-crHmcD eUERYU"
              data-testid="noRepeat"
              defaultChecked={true}
              name="schedule"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="noRepeat"
            />
            <label
              className="sc-egiyK gznvPK"
              htmlFor="schedule"
            >
              No repeat
            </label>
          </div>
          <div
            className="sc-bqiRlB kSktzV"
          >
            <input
              className="sc-crHmcD eUERYU"
              data-testid="specificDate"
              name="schedule"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="specificDate"
            />
            <label
              className="sc-egiyK gznvPK"
              htmlFor="schedule"
            >
              Specific date
            </label>
          </div>
          <div
            className="sc-bqiRlB kSktzV"
          >
            <input
              className="sc-crHmcD eUERYU"
              data-testid="daily"
              name="schedule"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="daily"
            />
            <label
              className="sc-egiyK gznvPK"
              htmlFor="schedule"
            >
              Daily
            </label>
          </div>
          <div
            className="sc-bqiRlB kSktzV"
          >
            <input
              className="sc-crHmcD eUERYU"
              data-testid="weekly"
              name="schedule"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="weekly"
            />
            <label
              className="sc-egiyK gznvPK"
              htmlFor="schedule"
            >
              Weekly
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-eCImPb jJXhll"
    >
      
      
      <button
        className="sc-jRQBWg grNFPM"
      >
        Cancel
      </button>
      <button
        className="sc-jRQBWg grNFPM"
        data-testid="submitForm"
      >
        OK
      </button>
    </div>
  </form>
</div>
`;
